### ISSUE TITLE
[Feature request] Implement episode groups

### ISSUE THREAD

**Issue by juanma-cvega** [jurialmunkey/plugin.video.themoviedb.helper#170]
I don't know whether you already knew (I just realised myself) but TMDB gives you the option to select grouping options for episodes. I just checked and it seems it's not available for every TV show. The ones I found the option in are animes. So apparently there is a way to change the default behaviour where some TV shows have a flat structure to split them into arcs. Each TV show has its own list of episode groups so it's not something you can set as an option for every show which makes implementing an episode group selector in the addon more complicated. 
So, is this something you've seen before and have thought about already?
Here is a [screenshot](https://pasteboard.co/J44qu9s.png).

**Comment by jurialmunkey** [comment]
No, I didn't actually know about this. I have seen the option for episode grouping in the API docs before but I didn't realise that this was what it does.

As you say, since there isn't a way to apply a blanket rule, it does make it significantly more difficult to implement.

It could be done if the default action is set to Browse so that you get the details list between Tvshow and Seasons. Then I could provide different episode grouping options at that level - that's the only way I could see it working. It would also mean it only works for tvshows -- it wouldn't work for episode based lists like in-progress or next aired.

**Comment by juanma-cvega** [comment]
> It could be done if the default action is set to Browse so that you get the details list between Tvshow and Seasons.

I had an idea but I think it's quite skin specific. In your skin horizon, you added the list of seasons as a submenu once inside a TV show. Could it be added there as an option? By default it could just use the default one but then give the option to select a different one when available. For other skins, it could be an option in the context menu once inside the list of episodes. 
The problem with this is that the option selected would need to be saved so users wouldn't need to change each time. And that once you are already in a season and you change episode group, it would need to either fallback to the first season or figure out the right season from the current episode selected. It could also affect library integration I guess. I don't know if episodes are named differently depending on the episode group.

**Comment by juanma-cvega** [comment]
To add to this, apparently there are 7 specific groups, so it's not specific to the TV show. The groups are these:
```
Original air date
Absolute
DVD
Digital
Story arc
Production
TV
```
What is specific to the TV show is the groups the show actually supports. So maybe a simple implementation could be a sorted list of preferred groups with a fallback strategy in the configuration settings? 
There is the question of how that would work with library integration as I have no idea if Kodi would support it.

**Comment by jurialmunkey** [comment]
At long last I've implemented this option - you can access episode groups from
`Tv Show > Context Menu > Browse lists > Episode groups`

I tested with Dragon Ball and it appears that the episodes still retain their season number that is on TMDb despite being split into groups (e.g. Third Dragonball saga "Red Ribbon Army Saga" appears as its own folder but contains episodes 1x29 to 1x45 rather than 3x1 to 3x16), so I don't think it will help with TVDb/Trakt matching. Even if I manually changed season numbers based on positioning (e.g. just assume group 3 is season 3), I think we'd end up with other issues pulling down the details from TMDb itself.

Either way, it's a cool feature that allows you to see air order and groupings based upon region etc.

**Comment by juanma-cvega** [comment]
Awesome! Thanks! I'll test once I have a little bit of time. 

> I tested with Dragon Ball and it appears that the episodes still retain their season number that is on TMDb despite being split into groups (e.g. Third Dragonball saga "Red Ribbon Army Saga" appears as its own folder but contains episodes 1x29 to 1x45 rather than 3x1 to 3x16), so I don't think it will help with TVDb/Trakt matching

I would expect that, yes. Even if you were able to do it, anime TV shows could follow a different grouping on TVDB so it wouldn't be worth it. It would help in some situations but not in all of them. The only reliable way to do it would be to allow TVDB as a source for the scraper but, as you said back in the day, that would mean rewriting the addon. And it wouldn't be called TMDB helper I guess :D 
Question about the new feature, is there any way to leave a default option for all TV shows? I remember when I looked into it that there were a predefined list of groupings to choose from. Not all TV shows had all of them so maybe a grouping preference list could be an option?

**Comment by jurialmunkey** [comment]
> The only reliable way to do it would be to allow TVDB as a source for the scraper but, as you said back in the day, that would mean rewriting the addon. And it wouldn't be called TMDB helper I guess :D

Well... The main issue is that from 2021 TVDb will become a paid service where either end users will need to pay a subscription fee or developers will need to buy a commercial API license. There's no point developing open source software that relies on TVDb. https://thetvdb.com/subscribe


> Question about the new feature, is there any way to leave a default option for all TV shows? I remember when I looked into it that there were a predefined list of groupings to choose from. Not all TV shows had all of them so maybe a grouping preference list could be an option?

Currently you can only view the episode groups via the Context Menu.

It isn't as if there is only one group per type, so I don't think there is much use to setting a default. For instance, Dragon Ball has these groups:
German Sagas (Type 5)
English Sagas (Type 5)
French Sagas (Type 5)
Toei's Official Divisions (Type 5)
Spanish Sagas (Type 5)
German TV (Type 7)

It's more that the episode groups relate to the different ways the show was released and the type just indicates what type of release that was. For example, Bleach has a CrunchyRoll group (Type 4) and a TV Tokyo group (Type 5). Totally possible for a show to have multiple releases of a particular type (e.g. different DVD editions that group episodes differently, or different streaming services, tv networks etc.)

**Comment by juanma-cvega** [comment]
> The main issue is that from 2021 TVDb will become a paid service where either end users will need to pay a subscription fee or developers will need to buy a commercial API license

I didn't know that. In that case it makes sense not to use TVDB but I wonder if that's gonna hit other addons/services that rely on it. Do you know if the addon is going to be impacted by this somehow? I think external IDs are populated by TMDB using a scraper but I guess it wouldn't be much for the service to pay the 12$ that costs the annual subscription. 

> It isn't as if there is only one group per type, so I don't think there is much use to setting a default

I didn't know that either. I thought there was a single group per type. 

I was just checking the groups on the TVDB site and they also have a DVD group and an absolute group but I'm not sure if that would help to do the matching from TMDB as the URLs don't seem to use it in the end. If you use absolute order you see all episodes on the website but then when you click on one, that's lost and you only get the episode ID in the URL, not the grouping selected. From the API perspective, maybe there is an endpoint that returns the list of episodes in one go which could then be used to do the matching but given what you said earlier about paying I guess that kind of API call wouldn't be for free.

**Comment by jurialmunkey** [comment]
> I didn't know that. In that case it makes sense not to use TVDB but I wonder if that's gonna hit other addons/services that rely on it. Do you know if the addon is going to be impacted by this somehow? I think external IDs are populated by TMDB using a scraper but I guess it wouldn't be much for the service to pay the 12$ that costs the annual subscription.

Yep, once 2021 rolls around it will impact all add-ons that use TVDb API. I doubt add-on authors are going to pay for a commercial license, so their users will need to pay the subscription fee to TVDb for it to work.

TheMovieDb used to have a bot that scraped TVDb but they stopped doing that quite a while ago (linking the TVDb IDs is now done manually by users), so it won't affect them at all. If anything, I think it will be good for TheMovieDb because I expect many long time users of TVDb will finally jump ship, which will mean more users dedicated to updating tvshow data on TheMovieDb instead.

It might affect Trakt because I think they do automated pull downs from both sites. Though, once again, ultimately probably for the better in terms of TheMovieDb data because there will be more motivation for users to update it instead of TVDb.

> Do you know if the addon is going to be impacted by this somehow? I think external IDs are populated by TMDB using a scraper but I guess it wouldn't be much for the service to pay the 12$ that costs the annual subscription.

Commercial developer license will be much much more expensive than $12... That's the _user_ subscription price _per_ user for apps that don't want to negotiate a commercial license (instead they get a free key but each user needs to subscribe individually then enter their personal key into the app -- much in the same way that OMDb works).

> I didn't know that either. I thought there was a single group per type.

Yeah I honestly thought the same. It wasn't until I started actually putting it into place that I realised that wasn't the case - My original thinking was the same as yours: have a setting to choose a "preferred" type and display that grouping for seasons if available but fallback to standard seasons view in other scenarios.

> I was just checking the groups on the TVDB site and they also have a DVD group and an absolute group but I'm not sure if that would help to do the matching from TMDB as the URLs don't seem to use it in the end.

I think which shows actually have these groupings is going to be too patchy anyway. Those types only seem to be added if there is a specific reason for them. For instance, "Freaks and Geeks" has a DVD and absolute order because the DVD order differed from the production order which also differed from the airing order on TV. 

For the most part though, tv shows don't have any groupings at all (e.g. massively popular shows like Breaking Bad or The Walking Dead don't have any episode groupings). Also, a lot that have separate groupings still won't have absolute or DVD (e.g. Dragon Ball has multiple type 5 and one type 7 but no absolute order or DVD order groups).
